source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst

```diff
@@ -1514,7 +1514,11 @@ servers. For a "node is recovering" error, single-threaded clients MUST NOT
 mark the topology as "stale". If a node is recovering for some time, an
 immediate scan may not gain useful information.
 
-The client SHOULD clear its connection pool for the server.
+If the client is connected to server version 4.2 or higher,
+the client SHOULD keep any connections it has to the server open,
+and SHOULD NOT clear its connection pool for the server.
+If the client is connected to server version 4.0 or lower, the client
+SHOULD clear its connection pool to the server.
```

Comment:
Is this supposed to apply only when connected to replica sets or does it also include mongos?

Reply:
Replica sets only per the top of the document.

Reply:
This is the SDAM spec. My question is when we are connected to a 4.2 Mongos, should we reset the pool after a "not master" error?

Reply:
I am not too familiar with sharded clusters, but my impression is a not master in a SC should be treated the same way it would be in a simple replica set. Are there any issues with keeping the pool in a sharded cluster?

Reply:
> but my impression is a not master in a SC should be treated the same way it would be in a simple replica set.

Yeah I think I agree. PyMongo treats these errors the same for both.

I do have a different concern though. This section says we should not reset the pool after "not master" or "node is recovering" errors. "node is recovering" includes errors caused during **shutdown** (not stepdown), like `InterruptedAtShutdown` and `ShutdownInProgress`. I think we want those shutdown errors to reset the pool even on 4.2.

Reply:
If a node is shutting down, what is the point of creating a new connection to this node which will then return another InterruptedAtShutdown error? It's just as efficient (more efficient in fact, as I understand it) to receive that error over the existing connection.

Reply:
I don't understand what you mean by "creating a new connection". We're only talking about clearing (closing) the connections in the pool.

Reply:
If no operations are sent to the server in question, then does it matter whether the pool is cleared or not?

Reply:
Yes it does matter. Let's say that the primary shuts down and that during the shutdown the application sees a ShutdownInProgress error. If we do not clear the pool then the Pool contains N dead sockets. When the node restarts we will still have N dead sockets and the next attempt to use this server will fail with a connection error. 

If instead we do clear the pool after ShutdownInProgress, then the next attempt to use this server will succeed on a new connection.

Reply:
I see your point, however given that we do not currently have a definition of "an error caused by server shutdown" I would like to not add figuring this out to the step down tests. What would you suggest as changes to make in this PR?

Reply:
I think we should introduce a new class of errors called "shutdown error" which includes `InterruptedAtShutdown` and `ShutdownInProgress` as the server does: https://github.com/mongodb/mongo/blob/r4.1.8/src/mongo/base/error_codes.err#L333

The we can say that drivers MUST clear the connection pool after a "shutdown error" regardless of the server version.

Reply:
@xdg What do you think about adding the above proposal to this step down project?

Reply:
I think it's a good idea.